pkg/virt-handler/guest-stats-updater.go

```diff
@@ -0,0 +1,258 @@
+package virthandler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	virtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	"kubevirt.io/kubevirt/pkg/safepath"
+	"kubevirt.io/kubevirt/pkg/util"
+	virtcache "kubevirt.io/kubevirt/pkg/virt-handler/cache"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/stats"
+)
+
+const (
+	statsUpdateInterval = 1 * time.Minute
+)
+
+type GuestStatsController struct {
+	host                 string
+	virtShareDir         string
+	clientset            kubecli.KubevirtClient
+	vmiInformer          cache.SharedIndexInformer
+	launcherClient       virtcache.LauncherClientInfoByVMI
+	podIsolationDetector isolation.PodIsolationDetector
+	logger               *log.FilteredLogger
+}
+
+func NewGuestStatsController(host string, virtShareDir string, clientset kubecli.KubevirtClient, vmiInformer cache.SharedIndexInformer) *GuestStatsController {
+	return &GuestStatsController{
+		host:                 host,
+		virtShareDir:         virtShareDir,
+		clientset:            clientset,
+		vmiInformer:          vmiInformer,
+		launcherClient:       virtcache.LauncherClientInfoByVMI{},
+		podIsolationDetector: isolation.NewSocketBasedIsolationDetector(virtShareDir),
+		logger:               log.Logger("guest-stat-controller").With("node", host),
+	}
+}
+
+func (c *GuestStatsController) Run(stopCh <-chan struct{}) error {
+	cache.WaitForCacheSync(stopCh, c.vmiInformer.HasSynced)
+
+	go wait.Until(c.execute, statsUpdateInterval, stopCh)
```

Comment:
I think you can also lower this to 15 or 30 seconds if you do a conceptial change. I would recommend one conceptual change:

Wait for each VM individually "roughly" for the update period with randomization, to distribute the updates of all VMs over the period.
 
 If we don't do that, we end up with hammering on a bigger node the API with 300 requests in one go, then  one minute nothing, then we hammer on the apiserver again.
 
 250 VMs/15 is just ~17 requests per second, compared to one peak of 300 requests which will most certainly hit the rate-limiter or hurt the apiserver performance. 30 seconds would be  just 8 requests per seconds, which is definitely not a concern.

Reply:
Interesting!
Sounds like a great improvement to me. Thank you @rmohr!